runtime/termscounter_test.go

```diff
@@ -0,0 +1,139 @@
+package runtime_test
+
+import (
+	"testing"
+	"time"
+
+	"go.etcd.io/gofail/runtime"
+)
+
+var __fp_ExampleString *runtime.Failpoint = runtime.NewFailpoint("ExampleString") //nolint:stylecheck
+
+func TestTermsCounter(t *testing.T) {
+	testcases := []struct {
+		name      string
+		fp        string
+		desc      string
+		runbefore int
+		runafter  int
+		want      int
+	}{
+		{
+			name:     "Terms limit Failpoint",
+			fp:       "ExampleString",
+			desc:     `10*sleep(10)->1*return("abc")`,
+			runafter: 12,
+			want:     11,
+		},
+		{
+			name:      "Inbetween Enabling Failpoint",
+			fp:        "ExampleString",
+			desc:      `10*sleep(10)->1*return("abc")`,
```

Comment:
One interesting followup would be to add a `panic` term, so we know the behavior of counter if user of gofail recovers the panic.

Reply:
I added a panic, gofail does not recover from it. It fails the test. I haven't included the change in the update.


> --- FAIL: TestTermsCounter/Inbetween_Enabling_Failpoint2 (0.02s)
> panic: failpoint panic: ExampleString [recovered]
>         panic: failpoint panic: ExampleString
> 
> goroutine 24 [running]:
> testing.tRunner.func1.2({0x12a7000, 0xc0000a1440})
>         /usr/local/go/src/testing/testing.go:1396 +0x24e
> testing.tRunner.func1()
>         /usr/local/go/src/testing/testing.go:1399 +0x39f
> panic({0x12a7000, 0xc0000a1440})
>         /usr/local/go/src/runtime/panic.go:884 +0x212
> go.etcd.io/gofail/runtime.actPanic(0xc0000a2d40)
>         /Users/prasadc/dev/repo/gofail/p1/runtime/terms.go:326 +0xe7
> go.etcd.io/gofail/runtime.(*term).do(...)
>         /Users/prasadc/dev/repo/gofail/p1/runtime/terms.go:294
> go.etcd.io/gofail/runtime.(*terms).eval(0xc0000a6640)
>         /Users/prasadc/dev/repo/gofail/p1/runtime/terms.go:108 +0x109
> go.etcd.io/gofail/runtime.(*Failpoint).Acquire(0x0?)
>         /Users/prasadc/dev/repo/gofail/p1/runtime/failpoint.go:38 +0xa5
> go.etcd.io/gofail/runtime_test.exampleFunc()
>         /Users/prasadc/dev/repo/gofail/p1/runtime/termscounter_test.go:145 +0x25
> go.etcd.io/gofail/runtime_test.TestTermsCounter.func1(0xc000132340)
>         /Users/prasadc/dev/repo/gofail/p1/runtime/termscounter_test.go:78 +0x106
> testing.tRunner(0xc000132340, 0xc0000a1350)
>         /usr/local/go/src/testing/testing.go:1446 +0x10b
> created by testing.(*T).Run
>         /usr/local/go/src/testing/testing.go:1493 +0x35f
> exit status 2
> FAIL    go.etcd.io/gofail/runtime       0.497s

Reply:
gofail doesn't recover panics, but the client using the library might wrap the gofail code in recovery.